src/Avalonia.Base/Controls/ResourceDictionary.cs

```diff
@@ -241,12 +242,27 @@ public bool TryGetValue(object key, out object? value)
             {
                 if (value is DeferredItem deffered)
                 {
-                    _inner[key] = value = deffered.Factory(null) switch
+                    // Avoid simple reentrancy, which could commonly occur on redefining the resource.
+                    if (s_lastDeferredItem == deffered)
```

Comment:
Would it make more sense to check based on the key?

Reply:
Yes, I was thinking about key as an alternative solution